rclrs/src/parameter/service.rs

```diff
@@ -1,89 +1,264 @@
-    pub fn new(rcl_node_mtx: Arc<Mutex<rcl_node_t>>) -> Result<Self, RclrsError> {
-        let describe_parameters_service = Service::new(Arc::clone(&rcl_node_mtx),
-            "describe_parameters",
-            |req_id: &rmw_request_id_t, req: DescribeParameters_Request| {
-                DescribeParameters_Response {
-                    descriptors: seq![]
-                }
+    pub(crate) fn new(
+        node: &Node,
+        parameter_map: Arc<Mutex<ParameterMap>>,
+    ) -> Result<Self, RclrsError> {
+        let fqn = node.fully_qualified_name();
+        // TODO(luca) make sure it is OK to have an Arc instead of a Weak here and cleanup on
+        // destruction is made for the parameter map.
+        let map = parameter_map.clone();
+        let describe_parameters_service = node.create_service(
+            &(fqn.clone() + "/describe_parameters"),
+            move |_req_id: &rmw_request_id_t, req: DescribeParameters_Request| {
+                let map = map.lock().unwrap();
+                let descriptors = req
+                    .names
+                    .into_iter()
+                    .map(|name| {
+                        let name = name.to_cstr().to_str().ok()?;
+                        let storage = map.storage.get(name)?;
+                        let mut descriptor = match storage {
+                            ParameterStorage::Declared(storage) => {
+                                let (integer_range, floating_point_range) =
+                                    storage.options.ranges.to_descriptor_ranges();
+                                ParameterDescriptor {
+                                    name: name.into(),
+                                    type_: Default::default(),
+                                    description: storage.options.description.clone().into(),
+                                    additional_constraints: storage
+                                        .options
+                                        .constraints
+                                        .clone()
+                                        .into(),
+                                    dynamic_typing: matches!(storage.kind, ParameterKind::Dynamic),
+                                    read_only: matches!(storage.value, DeclaredValue::ReadOnly(_)),
+                                    floating_point_range,
+                                    integer_range,
+                                }
+                            }
+                            ParameterStorage::Undeclared(_) => ParameterDescriptor {
+                                name: name.into(),
+                                dynamic_typing: true,
+                                ..Default::default()
+                            },
+                        };
+                        descriptor.type_ = storage.to_parameter_type();
+                        Some(descriptor)
+                    })
+                    .collect::<Option<_>>()
```

Comment:
Similar to my comment on `get_parameter_types`, I don't think this will give us the correct behavior which [expects the sequence in the response to always have the same length as the sequence in the request](https://docs.ros2.org/galactic/api/rcl_interfaces/srv/DescribeParameters.html).

I think we can map the `None` cases to a ParameterDescriptor which is default-initialized except for its name, which should match the request string in the same place in the sequence.

Reply:
Same answer as `get_parameter_types`! https://github.com/ros2-rust/ros2_rust/pull/342#discussion_r1529627501. Specifically this is what happens when I run on a `rclcpp` example:

```
lucadv@ros:~$ ros2 service call /minimal_publisher/describe_parameters rcl_interfaces/srv/DescribeParameters "{names: [use_sim_time]}"       
requester: making request: rcl_interfaces.srv.DescribeParameters_Request(names=['use_sim_time'])

response:
rcl_interfaces.srv.DescribeParameters_Response(descriptors=[rcl_interfaces.msg.ParameterDescriptor(name='use_sim_time', type=1, description='', additional_constraints='', read_only=False, dynamic_typing=False, floating_point_range=[], integer_range=[])])

lucadv@ros:~$ ros2 service call /minimal_publisher/describe_parameters rcl_interfaces/srv/DescribeParameters "{names: [use_sim_time, hello]}"
requester: making request: rcl_interfaces.srv.DescribeParameters_Request(names=['use_sim_time', 'hello'])

response:
rcl_interfaces.srv.DescribeParameters_Response(descriptors=[])
```

But again same that it should be behaving differently if undeclared parameters are allowed and I'll be fixing this.